File: datastructureswithpython/Trees/BinarySearchTree.py
class BSTNode:
    def __init__(self, value):
        self.value = value
        self.rightChild = None
        self.leftChild = None


class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self, value):
        self.root = self.insertNode(self.root, value)

    def remove(self, value):
        self.root = self.removeNode(self.root, value)

    def minvalue(self):
        minNode = self.minValueNode(self.root)
        return minNode.value

    def inorder(self):
        self.inOrder(self.root)

    def postorder(self):
        self.postOrder(self.root)

    def preorder(self):
        self.preOrder(self.root)

    def height(self):
        return self.heightCount(self.root)

    def search(self,value):
        return self.searchNode(self.root,value)

    def insertNode(self, node, value):
        if node is None:
            return BSTNode(value)
        elif node.value > value:
            node.leftChild = self.insertNode(node.leftChild, value)
        elif node.value < value:
            node.rightChild = self.insertNode(node.rightChild, value)
        elif node.value == value:
            return node
        return node

    def minValueNode(self, node):
        while node.leftChild is not None:
            node = node.leftChild
        return node

    def removeNode(self, node, value):
        if node is None:
            return node
        if node.value == value:
            if node.leftChild is None:
                return node.rightChild
            elif node.rightChild is None:
                return node.leftChild
            else:
                minNode = self.minValueNode(node.rightChild)
                node.value = minNode.value
                node.rightChild = self.removeNode(node.rightChild, minNode.value)
        elif node.value > value:
            node.leftChild = self.removeNode(node.leftChild, value)
        elif node.value < value:
            node.rightChild = self.removeNode(node.rightChild, value)
        return node

    def inOrder(self, node):
        if node is not None:
            self.inOrder(node.leftChild)
            print(node.value)
            self.inOrder(node.rightChild)

    def postOrder(self, node):
        if node is not None:
            self.postOrder(node.leftChild)
            self.postOrder(node.rightChild)
            print(node.value)

    def preOrder(self, node):
        if node is not None:
            print(node.value)
            self.preOrder(node.leftChild)
            self.preOrder(node.rightChild)

    def heightCount(self, node):
        if node is not None:
            return 1 + max(self.heightCount(node.leftChild), self.heightCount(node.rightChild))
        else:
            return 0

    def searchNode(self,node,value):
        if node is not None:
            if node.value == value:
                return True
            elif node.value > value:
                return self.searchNode(node.leftChild,value)
            else:
                return self.searchNode(node.rightChild,value)
        return False

File: datastructureswithjava/trees/BinarySearchTree.java
package datastructureswithjava.trees;

import java.util.LinkedList;
import java.util.Queue;

class BSTNode<T> {
    T value;
    BSTNode<T> leftChild = null,rightChild=null;
    BSTNode(T value){
        this.value = value;
    }
}

public class BinarySearchTree <T extends Comparable<T>>{

    private BSTNode<T> root = null;

    void preOrder(){
        preOrder(root);
        System.out.println();
    }
    void inOrder(){
        inOrder(root);
        System.out.println();
    }
    void postOrder(){
        postOrder(root);
        System.out.println();
    }
    private void preOrder(BSTNode<T> node){
        if (node!=null) {
            System.out.print(node.value + " ");
            preOrder(node.leftChild);
            preOrder(node.rightChild);
        }
    }
    private void inOrder(BSTNode<T> node){
        if (node!=null) {
            inOrder(node.leftChild);
            System.out.print(node.value + " ");
            inOrder(node.rightChild);
        }
    }
    private void postOrder(BSTNode<T> node){
        if (node!=null) {
            postOrder(node.leftChild);
            postOrder(node.rightChild);
            System.out.print(node.value + " ");
        }
    }
    void levelOrder(){
        if (root!=null){
            Queue<BSTNode<T>> queue = new LinkedList<BSTNode<T>>();
            queue.add(root);
            while (queue.size()!=0){
                BSTNode<T> current = queue.peek();
                queue.remove();
                System.out.print(current.value + " ");
                if (current.leftChild != null) queue.add(current.leftChild);
                if (current.rightChild != null) queue.add(current.rightChild);
            }
            System.out.println();
        }
    }
    void insert(T value){
        root = insert(root,value);
    }
    private BSTNode<T> insert(BSTNode<T> node,T value){
        if (node==null)
            return new BSTNode<>(value);
        if (node.value == value)
            return node;
        else if(node.value.compareTo(value)>0)
            node.leftChild = insert(node.leftChild,value);
        else
            node.rightChild = insert(node.rightChild,value);
        return node;
    }
    int height(){
        return height(root);
    }
    private int height(BSTNode<T> node){
        if(node == null){
            return 0;
        }
        else {
            return 1 + Math.max(height(node.leftChild),height(node.rightChild));
        }
    }

    T min(){
        return min(root);
    }
    private T min(BSTNode<T> node){
        if(node.leftChild!=null) return min(node.leftChild);
        return node.value;
    }
    T max(){
        return max(root);
    }
    private T max(BSTNode<T> node){
        if(node.rightChild!=null) return min(node.rightChild);
        return node.value;
    }
    boolean search(T value){
        return search(root,value);
    }
    private boolean search(BSTNode<T> node,T value){
        if (node!=null){
            if(node.value == value)
                return true;
            return search(node.leftChild,value) || search(node.rightChild,value);
        }
        return false;
    }
    void delete(T value){
        if (search(value))
            delete(root,value);
        else
            System.out.println("Item Not Found");
    }
    private BSTNode<T> delete(BSTNode<T> node,T value){
        if(node!=null){
            if(node.value.compareTo(value)==0){
                if(node.rightChild == null && node.leftChild == null){
                    return null;
                }
                else if(node.leftChild != null && node.rightChild == null){
                    return node.leftChild;
                }
                else if(node.rightChild != null && node.leftChild == null){
                    return node.rightChild;
                }else{
                    T minValue = min(node.rightChild);
                    node.value = minValue;
                    node.rightChild = delete(node.rightChild,minValue);
                }
            }
            else if(node.value.compareTo(value) > 0)
                node.leftChild = delete(node.leftChild,value);
            else
                node.rightChild = delete(node.rightChild,value);
        }
        return node;
    }
}